course9/lecture_en.md

```diff
@@ -0,0 +1,273 @@
+
+```moonbit no-check
+fn make[T: Default]() -> Queue[T] { // `T` should support the `default` method.
+  { 
+    array: Array::make(5, T::default()), // The return type of `default` is `T`.
+    start: 0,  end: 0,  length: 0 
+  } 
+}
+```
+
+With bounds, we can timely detect errors caused by calling missing methods. In the following example, we are trying to build a queue of `BoxedInt`, which however, does not support the `default` method. Consequently, the compiler notifies us of this error.
+
+![](../pics/no_method.png)
+
+Using the same approach, we can reimplement the `insert` method for `Tree`. In particular, we need to demand that the type parameter `T` is bound by `Compare`. This enables us to compare two `T` values by calling `T::compare` directly, without requiring an additional parameter. The `compare` method must have been provided by `T`, as it is bound by `Compare`.
+
+```moonbit
+fn insert[T : Compare](tree : Tree[T], value : T) -> Tree[T] {
+  // Since `T` is bound by `Compare`, it should support the `compare` method.
+  match tree {
+    Empty => Node(value, Empty, Empty) 
+    Node(v, left, right) =>  
+      if T::compare(value, v) == 0 { // We can call `compare` here.
+        tree
+      } else if T::compare(value, v) < 0 { // We can call `compare` here.
+        Node(v, insert(left, value), right)
+      } else {
+        Node(v, left, insert(right, value))
+      }
+  }
+}
+```
+
+## Implementation of Traits
+
+Since the trait system of MoonBit is structured, to implement a trait, all we need to do is to define the corresponding methods using the syntax `fn <type>::<method>(...) -> ...`.
+
+In the following example, we define the `default` method for the `BoxedInt` type.
+
+```moonbit
+struct BoxedInt { value : Int }
+
+fn BoxedInt::default() -> BoxedInt {
+  { value : Int::default() }
+  // The default value can be defined by boxing the default value of `Int`.
+}
+```
+
+By defining the `default` method, the `Default` trait is now implemented. Therefore, we can construct a queue of `BoxedInt` using the `make` function, which internally invokes the `default` method of `BoxedInt`.
+
+```moonbit no-check
+fn init {
+  let array: Queue[BoxedInt] = make()
+}
+```
+
+## Method Chaining
+
+In addition to `<type>::<method>(<expr>, ...)`, we can as well call the method using `<expr>.<method>(...)`, given `<expr>` is of type `<type>`. When defining such methods, `<type>::` can also be omitted when the first parameter is named `self`.
+
+In this way, when we need to call such functions sequentially, we can avoid multiple nesting of method calls, thereby enhancing the readability of the code.
+
+```moonbit
+fn BoxedInt::plus_one(b: BoxedInt) -> BoxedInt {
+  { value : b.value + 1 }
+}
+fn plus_two(self: BoxedInt) -> BoxedInt {
+  // `<type>::` can be omitted when the first parameter is named `self`.
```

Comment:
```suggestion
// `<type>::` can be omitted when the first parameter is named `self`.
fn plus_two(self: BoxedInt) -> BoxedInt {  
```
The `<type>::` is omitted in the type signature, so it may be better if the order is adjusted